spring-r2dbc/src/main/java/org/springframework/r2dbc/core/BeanPropertyRowMapper.java

```diff
@@ -0,0 +1,436 @@
+import org.springframework.beans.NotWritablePropertyException;
+import org.springframework.beans.TypeConverter;
+import org.springframework.beans.TypeMismatchException;
+import org.springframework.core.convert.ConversionService;
+import org.springframework.core.convert.support.DefaultConversionService;
+import org.springframework.dao.DataRetrievalFailureException;
+import org.springframework.dao.InvalidDataAccessApiUsageException;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * Mapping {@code BiFunction} implementation that converts a row into a new instance
+ * of the specified mapped target class. The mapped target class must be a
+ * top-level class or {@code static} nested class, and it must have a default or
+ * no-arg constructor.
+ *
+ * <p>Column values are mapped based on matching the column name (as obtained from
+ * R2DBC meta-data) to public setters in the target class for the corresponding
+ * properties. The names are matched either directly or by transforming a name
+ * separating the parts with underscores to the same name using "camel" case.
+ *
+ * <p>Mapping is provided for properties in the target class for many common types &mdash;
+ * for example: String, boolean, Boolean, byte, Byte, short, Short, int, Integer,
+ * long, Long, float, Float, double, Double, BigDecimal, {@code java.util.Date}, etc.
+ *
+ * <p>To facilitate mapping between columns and properties that don't have matching
+ * names, try using column aliases in the SQL statement like
+ * {@code "select fname as first_name from customer"}, where {@code first_name}
+ * can be mapped to a {@code setFirstName(String)} method in the target class.
+ *
+ * <p>For a {@code NULL} value read from the database, an attempt will be made to
+ * call the corresponding setter method with {@code null}, but in the case of
+ * Java primitives this will result in a {@link TypeMismatchException} by default.
+ * To ignore {@code NULL} database values for all primitive properties in the
+ * target class, set the {@code primitivesDefaultedForNullValue} flag to
+ * {@code true}. See {@link #setPrimitivesDefaultedForNullValue(boolean)} for
+ * details.
+ *
+ * <p>If you need to map to a target class which has a <em>data class</em> constructor
+ * &mdash; for example, a Java {@code record} or a Kotlin {@code data} class &mdash;
+ * use {@link DataClassRowMapper} instead.
+ *
+ * <p>Please note that this class is designed to provide convenience rather than
+ * high performance. For best performance, consider using a custom mapping function
+ * implementation.
+ *
+ * @author Thomas Risberg
+ * @author Juergen Hoeller
+ * @author Sam Brannen
+ * @author Simon Baslé
+ * @since 6.1
+ * @param <T> the result type
+ * @see DataClassRowMapper
+ */
+public class BeanPropertyRowMapper<T> implements BiFunction<Row, RowMetadata, T> {
```

Comment:
It might even be sufficient to let this be `Function<Readable>` for mapping stored procedure outputs in addition to `Row`s.